File: LLMgeneratedCode/gemini-1.R
# Install necessary packages if you don't have them already
if(!requireNamespace("GEOquery", quietly = TRUE)){
  install.packages("GEOquery")
}
if(!requireNamespace("Biobase", quietly = TRUE)){
  install.packages("Biobase")
}
if(!requireNamespace("dplyr", quietly = TRUE)){
  install.packages("dplyr")
}
if(!requireNamespace("caret", quietly = TRUE)){
  install.packages("caret")
}
if(!requireNamespace("ggplot2", quietly = TRUE)){
  install.packages("ggplot2")
}

# Load necessary libraries
library(GEOquery)
library(Biobase)
library(dplyr)
library(caret)
library(ggplot2)

# 1. Download GSE149440 dataset
gse <- getGEO("GSE149440", GSEMatrix =TRUE, getGPL=FALSE)

if (length(gse) > 1) idx <- grep("GSE149440_series_matrix", names(gse))[1] else idx <- 1
gse <- gse[[idx]]

# 2. Extract expression data and metadata
expression_data <- exprs(gse)
metadata <- pData(gse)

# 3. Prepare data for training and testing

# Extract gestational age and train/test split information
metadata <- metadata %>%
  mutate(gestational_age = as.numeric(`gestational age:ch1`),
         train_set = `train:ch1`)

# Separate training and test samples based on 'train_set' variable
train_metadata <- metadata %>% filter(train_set == '1')
test_metadata <- metadata %>% filter(train_set == '0')

# Get sample names for training and testing
train_samples <- rownames(train_metadata)
test_samples <- rownames(test_metadata)

# Filter expression data for training and testing samples
train_expression <- expression_data[, train_samples, drop = FALSE]
test_expression <- expression_data[, test_samples, drop = FALSE]

# Ensure we have gestational age data and expression data for both sets
if(nrow(train_metadata) == 0 || ncol(train_expression) == 0 || nrow(test_metadata) == 0 || ncol(test_expression) == 0) {
  stop("Problem separating training and testing data. Check metadata 'train:ch1' variable.")
}

# 4. Train a prediction model (Linear Regression)

# Create training data frame with gestational age as target and gene expression as predictors
train_df <- data.frame(gestational_age = train_metadata$gestational_age, t(train_expression))

# Train a linear regression model using caret package for simplicity
# We will use all genes as predictors. In a real scenario, feature selection might be important.
model <- train(
  gestational_age ~ .,
  data = train_df,
  method = "lm" # Linear Regression
)

# 5. Apply the model on the test set

# Create test data frame in the same format as training data (without gestational_age for prediction)
test_df_predictors <- data.frame(t(test_expression))
colnames(test_df_predictors) <- colnames(train_df)[-1] # Ensure column names match training data

# Predict gestational age for the test set
predictions <- predict(model, newdata = test_df_predictors)

# Get actual gestational ages for the test set
actual_gestational_age_test <- test_metadata$gestational_age

# 6. Calculate Root Mean Squared Error (RMSE) on the test set
rmse_value <- RMSE(predictions, actual_gestational_age_test)
cat("Root Mean Squared Error (RMSE) on the test set:", rmse_value, "\n")

# 7. Generate a scatter plot of predicted vs actual gestational age for the test set

plot_data <- data.frame(actual = actual_gestational_age_test, predicted = predictions)

scatter_plot <- ggplot(plot_data, aes(x = actual, y = predicted)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, color = "red", linetype = "dashed") + # Add y=x line for reference
  xlab("Actual Gestational Age") +
  ylab("Predicted Gestational Age") +
  ggtitle("Predicted vs Actual Gestational Age on Test Set") +
  theme_minimal()

print(scatter_plot)

# Optional: Save the plot to a file
# ggsave("gestational_age_prediction_scatter_plot.png", plot = scatter_plot, width = 6, height = 6, units = "in")

cat("Script finished. RMSE printed and scatter plot generated.\n")

File: LLMgeneratedCode/gemini-1.py
import GEOparse
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt

# Download the dataset
geo_id = "GSE149440"
dest_dir = "./data/"
gse = GEOparse.get_GEO(geo=geo_id, destdir=dest_dir)

# Access expression data and metadata
expression_data_df = gse.pivot_samples('VALUE')
metadata_df = gse.phenotype_data

# Convert gestational age to numeric and handle potential errors
metadata_df['gestational age:ch1'] = pd.to_numeric(metadata_df['gestational age:ch1'], errors='coerce')

# Identify training and test samples
train_samples_metadata = metadata_df[metadata_df['train:ch1'] == '1']
test_samples_metadata = metadata_df[metadata_df['train:ch1'] == '0']

# Get sample names for train and test sets
train_sample_names = train_samples_metadata.index
test_sample_names = test_samples_metadata.index

# Filter expression data for train and test samples
train_expression_data = expression_data_df[train_sample_names]
test_expression_data = expression_data_df[test_sample_names]

# Align metadata and expression data and drop samples with NaN gestational age in train set
train_metadata_aligned = train_samples_metadata.loc[train_sample_names].dropna(subset=['gestational age:ch1'])
train_expression_data_aligned = train_expression_data[train_metadata_aligned.index]

# Prepare training data
X_train = train_expression_data_aligned.T # Transpose so genes are features and samples are rows
y_train = train_metadata_aligned['gestational age:ch1']

# Train the linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Prepare test data, aligning and handling potential NaN gestational age in test set (though not strictly needed for prediction, good for evaluation if present)
test_metadata_aligned = test_samples_metadata.loc[test_sample_names].dropna(subset=['gestational age:ch1'])
test_expression_data_aligned = test_expression_data[test_metadata_aligned.index] # Use aligned indices

X_test = test_expression_data_aligned.T # Transpose for test data as well
y_test_actual = test_metadata_aligned['gestational age:ch1']

# Predict gestational age for test set
y_test_predicted = model.predict(X_test)

# Evaluate the model: Root Mean Squared Error (RMSE)
rmse = np.sqrt(mean_squared_error(y_test_actual, y_test_predicted))
print(f"Root Mean Squared Error (RMSE) on the test set: {rmse:.4f}")

# Generate scatter plot of predicted vs actual gestational age
plt.figure(figsize=(8, 6))
plt.scatter(y_test_actual, y_test_predicted, alpha=0.6)
plt.xlabel("Actual Gestational Age")
plt.ylabel("Predicted Gestational Age")
plt.title("Predicted vs Actual Gestational Age (Test Set)")
plt.plot([min(y_test_actual), max(y_test_actual)], [min(y_test_actual), max(y_test_actual)], color='red', linestyle='--', linewidth=1) # Diagonal line for perfect prediction
plt.grid(True)
plt.savefig("gestational_age_prediction_scatter_plot.png") # Save the plot
plt.savefig("./out2/gemini-1.py.pdf", format="pdf")